### PR TITLE
[router]: Platform specific routes

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

This is testing the concept of making routes/layouts/apis be platform specific. This is just an initial draft, it will be fleshed out after the holiday period.

Goal is to land this in v3 behind feature flags and be stable by v4

# How

Rules:

* API routes cannot have platform extensions
* Layouts and routes can have extensions only if a non-extension version exists

For example, you have have `/page.web.tsx` only if `/page.tsx` also exists. This makes platform extensions an additive change and won't break concepts like universal deep linking and typed routes

API routes do not support platform extensions as they are independent of a platform.

# Implementation

> This PR only partial solves this issue, it will need a follow up PR for typed routes.

I have created a new refactored version of `getRoutes()`. All new functionality (and error messages) are feature flagged turned off, including the function itself.

When `unstable_platformExtensions` are enabled, it will group routes with the same name by specificity, choosing the most specific route.

The general logic behind `getRoutes` is unchanged. It is a two pass system that
1. Converts the context keys to a tree
2. Traverses the tree
    - Hoisting routes to nearest layout
    - Adds entryPoints
    - Selects the most specific route

The main difference between this implementation is the use of lookup tables and no intermediate data structures, which should speed up startup time for larger apps.

# Test Plan

There are two new tests files

- `getRoutes-comparision.test.web.ts`: Simply compares the new `getRoutes()` vs the old.
- `getRoutes-new.test.web.ts`: Tests for the new `getRoutes()`

# Future work

**Required**

- Typed routes
  - Typed Routes only care about the URL route, not the file it resolves to.
  - All routes need a non-platform fallback so all we need to do is exclude the platform specific routes from being processed
  
**Optional** 

This isn't part of platform routes, but could be a good task afterwards

- `routes-manifest`
- Static rendering manifest

`routes-manifest` and static rendering are conceptually similar and could be merged. Currently both are efficient with multiple redundant loops over the routes. 

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-router/src/global-state/getRoutes.ts [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+  unstable_improvedErrorMessages?: boolean;
+};
+
+type DirectoryNode = {
+  layout?: RouteNode[];
+  views: Map<string, RouteNode[]>;
+  subdirectories: Map<string, DirectoryNode>;
+};
+
+const validPlatforms = new Set(['android', 'ios', 'windows', 'osx', 'native', 'web']);
```
Comment (marklawlor) [inline packages/expo-router/src/global-state/getRoutes.ts:29]: Need to add other out of tree platforms (tv?)

File: packages/expo-router/src/global-state/getRoutes.ts [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+
+    // This is a file that should be ignored
+    if (meta.specificity < 0) {
+      continue;
+    }
+
+    const leaves: DirectoryNode[] = [];
+    for (const key of extrapolateGroups(meta.key)) {
+      let node = directory;
+      const subdirectoryParts = key.replace(meta.filename, '').split('/').filter(Boolean);
```
Comment (marklawlor) [inline packages/expo-router/src/global-state/getRoutes.ts:75]: Think there is a cleaner method to get the parts

File: packages/expo-router/src/global-state/getRoutes.ts [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+        if (existing) {
+          throw new Error(
+            `The layouts "${filePath}" and ${existing.contextKey} conflict in "${meta.dirname}. Please remove one of these files.`
+          );
+        } else {
+          leaf.layout[meta.specificity] = node;
+        }
+      }
+    } else if (meta.isApi) {
+      // TODO
```
Comment (marklawlor) [inline packages/expo-router/src/global-state/getRoutes.ts:126]: TODO

File: packages/expo-router/src/global-state/getRoutes.ts [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+    if (meta.isLayout) {
+      hasLayout ||= leaves.length > 0;
+      for (const leaf of leaves) {
+        leaf.layout ??= [];
+
+        const existing = leaf.layout[meta.specificity];
+
+        if (existing) {
+          throw new Error(
```
Comment (marklawlor) [inline packages/expo-router/src/global-state/getRoutes.ts:118]: Needs test

File: packages/expo-router/src/global-state/getRoutes.ts [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+
+function getIgnoreList(options?: Options) {
+  const ignore: RegExp[] = [/^\.\/\+html\.[tj]sx?$/, ...(options?.ignore ?? [])];
+  if (options?.preserveApiRoutes !== true) {
+    ignore.push(/\+api\.[tj]sx?$/);
+  }
+  return ignore;
+}
+
+function extrapolateGroups(key: string, keys: Set<string> = new Set()): Set<string> {
```
Comment (marklawlor) [inline packages/expo-router/src/global-state/getRoutes.ts:368]: This should have tests

File: packages/expo-router/src/global-state/getRoutes.ts (hunk 1 above)
Comment (EvanBacon) [reply packages/expo-router/src/global-state/getRoutes.ts:29]: Should just have ios, android, web, and native.

**Comment by marklawlor** [comment]
I'm going to re-implement this as a series of PRs

**Pull request closed without merging**